### PR TITLE
2.4 Backport For LXD Constraints and Image Types

### PR DESCRIPTION
**Pull request by manadart**
## Description of change

This is a backport of functionality added to the development branch, for supporting constraints and image types when provisioning LXD container machines.

It cherry picks all commits from these PRs:
- https://github.com/juju/juju/pull/8825
- https://github.com/juju/juju/pull/8838
- https://github.com/juju/juju/pull/8869
- https://github.com/juju/juju/pull/8891

## QA steps

See: https://github.com/juju/juju/pull/8891

## Documentation changes

Yes.

## Bug reference

N/A

### CODE REVIEW

**Comment by manadart** [comment]
[Backport LXD Constraints Support to 2.4 Branch](https://trello.com/c/zIq2VA66/332-backport-lxd-constraints-support-to-24-branch)

File: container/lxd/container.go [hunk 1]
```diff
@@ -181,8 +202,11 @@ func (s *Server) ContainerAddresses(name string) ([]network.Address, error) {
 // If the container fails to be started, it is removed.
 // Upon successful creation and start, the container is returned.
 func (s *Server) CreateContainerFromSpec(spec ContainerSpec) (*Container, error) {
+	logger.Infof("starting container %q (image %q)...", spec.Name, spec.Image.Image.Filename)
```
Comment (SimonRichardson) [inline container/lxd/container.go]: Are we going to put any additional logging here like with 2.5?
- https://bugs.launchpad.net/juju/+bug/1779897

Comment (manadart) [reply container/lxd/container.go]: Yeah, I'll remember to make that change here for later forward-porting.

**Comment by manadart** [comment]
$$merge$$

**Pull request merged**